### PR TITLE
Refactor some code into shared library

### PR DESCRIPTION
**Pull request by JohnMcPMS**
## Change
Refactor some code into a static library for use in future work.  The main goal in the separation is to ensure that the winget specifics (settings, telemetry, etc.) aren't brought in accidentally.

The biggest code changes required were:
1. Moving the logger helpers into their respective owning types
2. Refactoring `ThreadGlobals` into an interface that can be implemented by the using component

Everything else should be changes to account for that.

## Validation
Unit tests and manual validation of the command line.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/winget-cli/pull/2844)

### CODE REVIEW

File: src/AppInstallerSharedLib/YamlWrapper.h [hunk 1]
```diff
@@ -0,0 +1,168 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (florelis) [inline src/AppInstallerSharedLib/YamlWrapper.h:1]: This file is added instead of moved?

**Review by florelis — APPROVED** [review]

File: src/AppInstallerSharedLib/YamlWrapper.h (hunk 1 above)
Comment (JohnMcPMS) [reply src/AppInstallerSharedLib/YamlWrapper.h:1]: Looks like I forgot to delete it in CommonCore...

File: src/AppInstallerSharedLib/AppInstallerSharedLib.vcxproj [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)'=='Fuzzing'">
+    <ClCompile>
+      <Optimization>MaxSpeed</Optimization>
+      <FunctionLevelLinking>true</FunctionLevelLinking>
+      <IntrinsicFunctions>true</IntrinsicFunctions>
+      <PreprocessorDefinitions>_NO_ASYNCRTIMP;NDEBUG;%(PreprocessorDefinitions);CLICOREDLLBUILD;WINGET_DISABLE_FOR_FUZZING</PreprocessorDefinitions>
+      <AdditionalIncludeDirectories>$(ProjectDir);$(ProjectDir)Public;$(ProjectDir)Telemetry;$(ProjectDir)..\binver;$(ProjectDir)..\YamlCppLib\libyaml\include;$(ProjectDir)..\cpprestsdk\cpprestsdk\Release\include;$(ProjectDir)..\JsonCppLib;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
+      <TreatWarningAsError>true</TreatWarningAsError>
+      <LanguageStandard>stdcpp17</LanguageStandard>
```
Comment (yao-msft) [inline src/AppInstallerSharedLib/AppInstallerSharedLib.vcxproj:250]: ><LanguageStandard>stdcpp17</LanguageStandard> [](http://example.com/codeflow?start=6&length=45)

language standard stdcpp17 should be applied to Debug and Release configurations above too

**Review by yao-msft — APPROVED** [review]
:shipit:

File: src/AppInstallerSharedLib/AppInstallerSharedLib.vcxproj (hunk 2 above)
Comment (JohnMcPMS) [reply src/AppInstallerSharedLib/AppInstallerSharedLib.vcxproj:250]: The project file was copied directly from CommonCore, and VS 2022 is showing that 17 is in effect despite not being set in either place.  🤷

**Pull request merged**